### PR TITLE
Add definedBy property for authenticators.

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Task Issue:
- https://github.com/wso2/product-is/issues/21111

Related PRs:
- https://github.com/wso2/carbon-identity-framework/pull/5991
- https://github.com/wso2/identity-api-server/pull/679

With this following properties will be introduce to the authenticator configs and Application Authenticators.
- `DefinedBy` - to indicate whether the authenticator is system-defined or user-defined. 
1. SYSTEM: Sytem defined authenticators, all the existing authenticators  are this type.
2. USER: User defined authenticators, all the custom authenticators of authentication extension come under this type.

With this PR following changes are added.

1. Add new property for authenticator config classes to hold the authenticator definedBy type.
2. Improve service layer with this property.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java [hunk 1]
```diff
@@ -101,9 +108,17 @@ public static FederatedAuthenticatorConfig build(OMElement federatedAuthenticato
...
+                federatedAuthenticatorConfig.setDefinedByType(DefinedByType.valueOf(element.getText()));
             }
         }
 
+        if (federatedAuthenticatorConfig.getDefinedByType() == null) {
+            federatedAuthenticatorConfig.setDefinedByType(DefinedByType.SYSTEM);
+            LOG.debug("The defined by type is not set for the {}. Hence setting default SYSTEM value.",
+                    federatedAuthenticatorConfig.getName());
+        }
+
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java:121]: Ideally IMO, this is redundant.
We need to make sure from where we load the auth config and those points are what should be fixed

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java [hunk 2]
```diff
@@ -111,8 +118,17 @@ public static LocalAuthenticatorConfig build(OMElement localAuthenticatorConfigO
...
+                localAuthenticatorConfig.setDefinedByType(DefinedByType.valueOf(member.getText()));
             }
         }
+
+        if (localAuthenticatorConfig.getDefinedByType() == null) {
+            localAuthenticatorConfig.setDefinedByType(DefinedByType.SYSTEM);
+            LOG.debug("The defined by type is not set for the {}. Hence setting default SYSTEM value.",
+                    localAuthenticatorConfig.getName());
+        }
+
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java:131]: IMO this is redundant

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java [hunk 3]
```diff
@@ -508,13 +509,15 @@ protected void setAuthenticator(ApplicationAuthenticator authenticator) {
             localAuthenticatorConfig.setTags(getTags(authenticator));
             AuthenticatorConfig fileBasedConfig = getAuthenticatorConfig(authenticator.getName());
             localAuthenticatorConfig.setEnabled(fileBasedConfig.isEnabled());
+            localAuthenticatorConfig.setDefinedByType(DefinedByType.SYSTEM);
```
Comment (malithie) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java:512]: If we are giving this property to authenticator then we should get it from there. else it's fine to say any authenticator registers as an osgi is a system authenticator

File: components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/AuthenticatorPropertiesConstant.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.base;
+
+public class AuthenticatorPropertiesConstant {
```
Comment (malithie) [inline components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/AuthenticatorPropertiesConstant.java:21]: AuthenticatorPropertyConstants

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java (hunk 1 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java]: Addressed with commit fe95d4b60aeb8557a6ac7cfe51ce2576ee94112c

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java (hunk 2 above)
Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/LocalAuthenticatorConfig.java]: Addressed with commit fe95d4b60aeb8557a6ac7cfe51ce2576ee94112c

File: components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/AuthenticatorPropertiesConstant.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/AuthenticatorPropertiesConstant.java]: Addressed with commit fe95d4b60aeb8557a6ac7cfe51ce2576ee94112c

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java (hunk 3 above)
Comment (Thisara-Welmilla) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java:512]: Addressed with commit fe95d4b60aeb8557a6ac7cfe51ce2576ee94112c. Only Authenticator config has this property now.

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java [hunk 5]
```diff
@@ -46,6 +49,7 @@
 public class FederatedAuthenticatorConfig implements Serializable {
 
     private static final long serialVersionUID = -2361107623257323257L;
+    private static final Logger LOG = LoggerFactory.getLogger(LocalAuthenticatorConfig.class);
```
Comment (pamodaaw) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java:52]:
```suggestion
    private static final Logger LOG = LoggerFactory.getLogger(FederatedAuthenticatorConfig.class);
```

Comment (Thisara-Welmilla) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/FederatedAuthenticatorConfig.java]: Addressed

**Review by malithie — APPROVED** [review]

**Pull request merged**